workflow/workflow_test.go

```diff
@@ -2111,6 +2111,402 @@ outputs:
+		"step_enabled": false,
+	})
+	assert.NoError(t, err)
+	assert.Equals(t, outputID, "workflow-success")
+	assert.Equals(t, outputData.(map[any]any), map[any]any{})
+}
+
+var optionalFieldWithFailureStepWorkflow = `
+version: v0.2.0
+input:
+  root: WorkflowInput
+  objects:
+    WorkflowInput:
+      id: WorkflowInput
+      properties:
+        fail_purposefully:
+          type:
+            type_id: bool
+steps:
+  hello_step:
+    plugin:
+      src: "n/a"
+      deployment_type: "builtin"
+    step: hello
+    input:
+      fail: !expr $.input.fail_purposefully
+  wait_step:
+    plugin:
+      src: "n/a"
+      deployment_type: "builtin"
+    step: wait
+    input:
+      wait_time_ms: 0
+outputs:
+  workflow-success:
+    hello_output: !wait-optional $.steps.hello_step.outputs.success
+    wait_output: !expr $.steps.wait_step.outputs.success
+`
+
+func TestOptionalFieldWithFailureStepWorkflow(t *testing.T) {
+	// Test a workflow where one of the steps can fail without
+	// causing a failure of the output.
+	preparedWorkflow := assert.NoErrorR[workflow.ExecutableWorkflow](t)(
+		getTestImplPreparedWorkflow(t, optionalFieldWithFailureStepWorkflow),
+	)
+	outputID, outputData, err := preparedWorkflow.Execute(context.Background(), map[string]any{
+		"fail_purposefully": false,
+	})
+	assert.NoError(t, err)
+	assert.Equals(t, outputID, "workflow-success")
+	assert.Equals(t, outputData.(map[any]any), map[any]any{
+		"wait_output": map[any]any{
+			"message": "Plugin slept for 0 ms.",
+		},
+		"hello_output": map[any]any{},
+	})
+	outputID, outputData, err = preparedWorkflow.Execute(context.Background(), map[string]any{
+		"fail_purposefully": true,
+	})
+	assert.NoError(t, err)
+	assert.Equals(t, outputID, "workflow-success")
+	assert.Equals(t, outputData.(map[any]any), map[any]any{
+		"wait_output": map[any]any{
+			"message": "Plugin slept for 0 ms.",
+		},
+	})
+}
+
+var softOptionalFieldWorkflow = `
+version: v0.2.0
+input:
+  root: WorkflowInput
+  objects:
+    WorkflowInput:
+      id: WorkflowInput
+      properties:
+        wait_1_ms:
+          type:
+            type_id: integer
+        wait_2_ms:
+          type:
+            type_id: integer
+        wait_1_enabled:
+          type:
+            type_id: bool
+        wait_2_enabled:
+          type:
+            type_id: bool
+steps:
+  wait_step_1:
+    plugin:
+      src: "n/a"
+      deployment_type: "builtin"
+    step: wait
+    input:
+      wait_time_ms: !expr $.input.wait_1_ms
+    enabled: !expr $.input.wait_1_enabled
+  wait_step_2:
+    plugin:
+      src: "n/a"
+      deployment_type: "builtin"
+    step: wait
+    input:
+      wait_time_ms: !expr $.input.wait_2_ms
+    enabled: !expr $.input.wait_2_enabled
+outputs:
+  workflow-success:
+    required_output: !expr $.steps.wait_step_1.outputs.success
+    optional_output: !soft-optional $.steps.wait_step_2.outputs.success
+`
+
+func TestSoftOptionalFieldWorkflow(t *testing.T) {
```

Comment:
Can you explain why this test needs to use a comparison of time delays to make assertions about the source code behavior, and if possible add a succinct version of that in the test's header comment?

Reply:
My goal was to isolate the soft-optional dependency as much as possible. Soft-optional provides no guarantees for ordering, so the design of this test is to use a large enough time to make it all but guaranteed that the intended outputs will be included.